### PR TITLE
All: Handle link in automatic title generation

### LINKED ISSUE #2915: Attaching an image or a file to a new empty note puts markdown tag into the title
Steps to reproduce:

1. Create a new empty note
2. Attach any file or image
3. Save the note

Maybe set the title from the file name instead or just "Untitled"?

### PR DESCRIPTION
**Pull request by anihm136**
Fixes #2915 

When the first line of a new note matches the regex for a markdown link, only the file name is taken as the title

Screenshot:
mdlink

### CODE REVIEW

File: ReactNativeClient/lib/models/Note.js [hunk 1]
```diff
@@ -85,13 +85,19 @@ class Note extends BaseItem {
...
 			const lines = body.trim().split('\n');
 			let output = lines[0].trim();
-			// Remove the first #, *, etc.
-			while (output.length) {
-				const c = output[0];
-				if (['#', ' ', '\n', '\t', '*', '`', '-'].indexOf(c) >= 0) {
-					output = output.substr(1);
-				} else {
-					break;
+			if (output.match(/\[.+\]\(.+\)/)) {
```
Comment (roman-r-m) [inline ReactNativeClient/lib/models/Note.js]: I think this is not going to work for images where tag starts with !

File: ReactNativeClient/lib/models/Note.js [hunk 2]
```diff
@@ -85,13 +85,19 @@ class Note extends BaseItem {
...
+				const start = output.indexOf('[') + 1;
+				const stop = output.indexOf(']') - 1;
```
Comment (roman-r-m) [inline ReactNativeClient/lib/models/Note.js]: Maybe I'm nitpciking here but the 2nd parameter of substr is not the end index but length.
In this case the result would always be the same thanks to the trim call above.

**Comment by roman-r-m** [comment]
Since I was the one who raised that issue..
I think it would be nice to handle the case where there's other text before of after the tag, like this:
```
Look I'm attaching [file.txt](:/11111) -> Look I'm attaching file.txt
```
But I don't really want this to get too complicated.

File: ReactNativeClient/lib/models/Note.js (hunk 1 above)
Comment (anihm136) [reply ReactNativeClient/lib/models/Note.js]: Yes, I'll just handle that case momentarily

**Comment by anihm136** [comment]
> I think it would be nice to handle the case where there's other text before of after the tag, like this:

This certainly makes sense. I'll try implementing this now

File: ReactNativeClient/lib/models/Note.js (hunk 2 above)
Comment (anihm136) [reply ReactNativeClient/lib/models/Note.js]: Ah, thanks for pointing that out! Using the wrong function

**Comment by anihm136** [comment]
> Since I was the one who raised that issue..
I think it would be nice to handle the case where there's other text before of after the tag, like this:
```Look I'm attaching [file.txt](:/11111) -> Look I'm attaching file.txt```
But I don't really want this to get too complicated.

Added some changes which should implement this, and handle image links as well. @roman-r-m could you take a look and see if this is what you're expecting?
mdlink2

**Comment by roman-r-m** [comment]
Looks good to me (though I do not have permissions to merge, have to wait for someone else to confirm).

One (admittedly artificial) example where the regepx will fail is something like
```
[Test] [file.txt](:/123456)
or
[file.txt](:/123456) (test)
```
But I think it's unlikely to happen in practice.

File: ReactNativeClient/lib/models/Note.js [hunk 3]
```diff
@@ -85,13 +85,18 @@ class Note extends BaseItem {
...
 			const lines = body.trim().split('\n');
 			let output = lines[0].trim();
-			// Remove the first #, *, etc.
-			while (output.length) {
-				const c = output[0];
-				if (['#', ' ', '\n', '\t', '*', '`', '-'].indexOf(c) >= 0) {
-					output = output.substr(1);
-				} else {
-					break;
+			const mdLinkRegex = /!?\[(.+)\]\(.+\)/;
```
Comment (laurent22) [inline ReactNativeClient/lib/models/Note.js:88]: What if the image markup is inside the body of the need note?

**Comment by laurent22** [comment]
We'll need test units for this change. Additionally there are many helper functions in MarkdownUtils, including many that deal with image markup, could something be reused from there?

File: ReactNativeClient/lib/models/Note.js (hunk 3 above)
Comment (anihm136) [reply ReactNativeClient/lib/models/Note.js:88]: So I would have to check for the image markup after removing the leading characters right?

**Comment by anihm136** [comment]
I'll look into the functions in MarkdownUtils and also write the test units. While I'm at it, do you think it makes sense to remove all markdown formatting from the default title, or is that out of scope (or not required)?

**Comment by anihm136** [comment]
I went through markdownUtils, and none of the functions appear to do what we want here i.e, extract link description from link. Would it be better if I write the logic as a function in markdownUtils and use it from there? That way it would be easy to add to the existing tests for that class also

**Comment by laurent22** [comment]
Yes it would be better to put the whole logic in a function of markdownUtils, as it will also make it easier to test it.

File: ReactNativeClient/lib/models/Note.js [hunk 4]
```diff
@@ -85,21 +86,16 @@ class Note extends BaseItem {
...
+			while (output.length) {
+				const c = output[0];
+				if (['#', ' ', '\n', '\t', '*', '`', '-'].indexOf(c) >= 0) {
+					output = output.substr(1);
+				} else {
+					break;
 				}
 			}
-			return output.substr(0, 80).trim();
+			return mdUtils.filterLink(output).substr(0, 80).trim();
```
Comment (laurent22) [inline ReactNativeClient/lib/models/Note.js]: Why do you put half the logic in your function, and you leave the other half here? Wouldn't it be logical to move everything to the function?

File: CliClient/tests/markdownUtils.js [hunk 5]
```diff
@@ -66,4 +66,19 @@ describe('markdownUtils', function() {
 		}
 	}));
 
+	it('replace markdown link with description', asyncTest(async () => {
+
+		const testCases = [
+			['Test case [one](link)', 'Test case one'],
+			['Test case ![two](imagelink)', 'Test case two'],
+			['Test case three', 'Test case three'],
+		];
```
Comment (laurent22) [inline CliClient/tests/markdownUtils.js:75]: We need more tests: empty string, string longer than the limit, images with no title, multiple images. Then once you've added the missing logic to your function, we need tests for this too.

File: ReactNativeClient/lib/markdownUtils.js [hunk 6]
```diff
@@ -91,6 +91,15 @@ const markdownUtils = {
 
 		return output.join('\n');
 	},
+
+	filterLink(mdLink) {
```
Comment (laurent22) [inline ReactNativeClient/lib/markdownUtils.js:95]: Why call if "filterLink" when it doesn't even filter links? Maybe filterImage, but something more specific like "titleFromBody" would be better.

Comment (laurent22) [reply ReactNativeClient/lib/markdownUtils.js:95]: And why is the input called "mdLink". Does your function not work if the input is not a link?

File: ReactNativeClient/lib/models/Note.js (hunk 4 above)
Comment (anihm136) [reply ReactNativeClient/lib/models/Note.js]: Alright. So I'll make one function to filter out markdown from the title and use it here. Just one doubt, would it make sense to remove the `defaultTitleFromBody` function altogether and do the entire thing in `defaultTitle` itself?

File: CliClient/tests/markdownUtils.js (hunk 5 above)
Comment (anihm136) [reply CliClient/tests/markdownUtils.js:75]: Will do

**Comment by anihm136** [comment]
I've moved the entire logic to markdownUtils and added tests. Its current behavior is - 
* Removes all leading *,#,-,spaces,tabs,newlines
* Markdown links (and image links) with description provided are replaced with the description
* Links without description are replaced by the link target URL

File: ReactNativeClient/lib/markdownUtils.js (hunk 6 above)
Comment (anihm136) [reply ReactNativeClient/lib/markdownUtils.js]: Now it returns the cleaned title from the body

**Comment by laurent22** [comment]
In that pull request, we use remove-markdown which I guess could be used in this pull request too? https://github.com/laurent22/joplin/pull/3037

**Comment by anihm136** [comment]
Yeas, that looks like the right thing for this application. It can probably also be used for the note preview. I'll change accordingly

**Comment by anihm136** [comment]
I tried using remove-markdown. While its obviously easier to use, it does not fulfil all the requirements. Specifically:
1. It does not handle links without description (completely removes instead of showing url, as mentioned)
2. It does not handle regular text enclosed in []
If these can be overlooked, I can go ahead and make the change.

On a side note, what is the procedure to add a new module to Joplin? Just add it to package.json?

**Comment by Rishgod** [comment]
> On a side note, what is the procedure to add a new module to Joplin? Just add it to package.json?

Yes. here is a similar conversation https://discourse.joplinapp.org/t/pull-request-help/6357

**Comment by laurent22** [comment]
Thanks for reviewing the package, then indeed I don't think we can use it. We probably shouldn't use it in the other PR either.

**Comment by anihm136** [comment]
Thanks for the clarification @Rishgod !

@laurent22 Since this seems to be a feature required across the board, would it be useful to add a regex-based function to markdownUtils to handle removal of all markdown? This seems to be the approach taken by the package also, except that it is not complete (and does not seem to be currently maintained either)

**Comment by laurent22** [comment]
Parsing Markdown and HTML with regex is not the way to go as there will always be cases it doesn't work. Instead we should use the markdown-it parser and create a plugin that extract the plain text only.

**Comment by anihm136** [comment]
Alright, I'll try implementing it this way

**Comment by anihm136** [comment]
@laurent22 I've written a function to extract text from HTML using `markdown-it`. However, I'm not sure whether it should be a part of this PR, the note preview one ([here](https://github.com/laurent22/joplin/pull/2882)) or a separate one by itself.

Also, for this particular issue, I feel using regex would be easier because
1. There aren't too many corner cases to be handled
2. Certain features required here (such as showing link when link description isn't present) would be much harder to implement using the parser

Please let me know what you think about these, and I'll proceed accordingly

**Comment by laurent22** [comment]
Thanks @anihm136, your solution indeed makes sense, so let's merge. There was a few small issue and a conflict which I've fixed now.
